docs/design/pnc.md

```diff
@@ -0,0 +1,113 @@
+
+### PNC instance configuration
+Currently, the url of the PNC instance is configured as a part of OSBS config. 
+
+## Overview of the current implementation in OSBS
+
+[atomic_reactor/plugins/fetch_maven_artifacts.py](https://github.com/containerbuildsystem/atomic-reactor/blob/master/atomic_reactor/plugins/fetch_maven_artifacts.py)
+1) OSBS loads the `fetch-artifacts-pnc.yaml`
+2) Information about individual artifacts is fetched from PNC api at `/artifacts/{id}` endpoint
+   - `publicUrl` key is used to get download url
+   - supported checksums are saved
+   - build ids are saved
+3) Download queue is formed containing above info for each artifact
+4) SBOM component info for each artifact is fetched from PNC @ `/artifacts/{id}`, `purl` key
+5) Queue is executed and each artifact is downloaded, checksum verified
+6) SBOM is compiled, including PNC components from step 4)
+
+
+### Missing features
+
+PNC SBOM components don't have the most complete purl, e.g.:
+`pkg:maven/org.kie.trustyai/explainability-service@999.0.0.managedsvc-redhat-00001?type=zip`
+
+According to [purl spec](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#maven), `classfier` and
+`type` parameters should be also defined, method is currently pending depending on if the artifacts are single components,
+or products that contain other components. Purl should also specify a `repository_url` parameter.
+
+
+
+## Design for the implementation in Cachi2
+
+
+### Providing the content for the hermetic build
+
+#### PNC config
+Existing `fetch-artifacts-pnc.yaml` file ([docs](https://osbs.readthedocs.io/en/osbs_ocp3/users.html?highlight=fetch#fetch-artifacts-pnc-yaml))
+will be used as configuration. It needs to be extended to also include PNC instance url.
+
+#### Pre-fetch PNC dependencies
+Cachi2 will use the `artifacts` key specified in the `fetch-artifacts-pnc.yaml` file to fetch info about given artifacts
+from PNC (`/artifacts/{id}` endpoint), including url to download the artifact, checksums to verify integrity, and the purl.
+The artifacts should be fetched into a separate subdirectory in cachi2 output folder, and into the target path.
+
+##### Fetching sources
+Cachi2 will use the `build_id` keys to fetch build sources from PNC (`/builds/{id}/scm-archive` endpoint). These sources
+should be fetched into a separate subdirectory in cachi2 output folder, different from the artifacts.
+
+### Generating the SBOM
+As mentioned in [missing features](#missing-features), the purl needs to specify more parameters
+- `repository_url` - obtainable by using host part of `publicUrl` and the path in `targetRepository.repositoryPath`
+  from PNC `/artifacts/{id}` endpoint response.
+- `type` & `classifier` - solution pending
+
+#### Relevant configuration for the build
+No environment variables need to be set.
+
+#### Injecting files to an expected location
+The actual artifacts should be injected into `artifacts/` path in the repository. 
```

Comment:
It is Cachi2's philosophy to avoid modifying the source files at all costs. That's why we have a separate step after [fetch-deps](https://github.com/containerbuildsystem/cachi2/blob/main/docs/usage.md#pre-fetch-dependencies) called [inject-files](https://github.com/containerbuildsystem/cachi2/blob/main/docs/usage.md#inject-project-files). The latter is used to modify source files as needed to make the build work hermetically.

But if I understood this correctly, this is simply placing the prefetched dependencies in a specific folder, right? It doesn't quite seem to be the use case for `inject-files`.

Will maven/gradle/build tool necessarily look for the content in `./artifacts`? Is this configurable somehow?

Reply:
> Will maven/gradle/build tool necessarily look for the content in ./artifacts? Is this configurable somehow?

Do they look there at all? I kind of assumed ./artifacts was just a made up place where OSBS used to place the downloaded archives, and OSBS users then had to explicitly use the stuff from ./artifacts in their Dockerfile

Reply:
I think it's pretty arbitrary. I don't think it is 100% necessary, but on the other hand, it will ease the migration for those middleware source repos, which are the targeted customer of this change, as their Dockerfiles are already pointing at that directory. 

Let me know if this is an issue, or something that would go against how cachi2 is designed to work.

Reply:
Cachi2 always places the downloaded files under `{output-folder}/deps/{package_manager}`, so we should probably respect that. I don't think `inject-files` is meant to simply move these files to a new location either.

It feels to me this is out of scope for Cachi2, but since this is a specific problem for the Konflux use-case, we could potentially work around this in the Konflux pipeline.

Reply:
Agreed. Will edit so it is clear those files will **not** be in the `/artifacts` location, but instead will use the default cachi2 location.